dbusmock/templates/systemd-base.py

```diff
@@ -0,0 +1,91 @@
+'''systemd mock template base
+'''
+
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of the GNU Lesser General Public License as published by the Free
+# Software Foundation; either version 3 of the License, or (at your option) any
+# later version.  See http://www.gnu.org/copyleft/lgpl.html for the full text
+# of the license.
+
+__author__ = 'Jonas Ådahl'
+__copyright__ = '(c) 2021 Red Hat'
+
+from gi.repository import GLib
+
+BUS_NAME = 'org.freedesktop.systemd1'
+MAIN_OBJ = '/org/freedesktop/systemd1'
+MAIN_IFACE = 'org.freedesktop.systemd1.Manager'
+
+
+def load_base(mock, _parameters):
+    mock.next_job_id = 1
+    mock.units = {}
+
+    mock.AddMethod(MAIN_IFACE,
+                   'StartUnit',
+                   'ss', 'o',
+                   'ret = self.StartUnit(self, args[0], args[1])')
+    mock.AddMethod(MAIN_IFACE,
+                   'StartTransientUnit',
+                   'ssa(sv)a(sa(sv))', 'o',
+                   'ret = self.StartTransientUnit(self, args[0], args[1], args[2], args[3])')
+    mock.AddMethod(MAIN_IFACE,
+                   'StopUnit',
+                   'ss', 'o',
+                   'ret = self.StopUnit(self, args[0], args[1])')
+
+    mock.StartUnit = StartUnit
+    mock.StartTransientUnit = StartTransientUnit
+    mock.StopUnit = StopUnit
+
+
+def escape_unit_name(name):
+    for s in ['.', '-']:
+        name = name.replace(s, '_')
+
+
+def emit_job_new_remove(self, job_id, job_path, name):
+    self.EmitSignal('org.freedesktop.systemd1.Manager', 'JobNew', 'uos',
+                    [job_id, job_path, name])
+    self.EmitSignal('org.freedesktop.systemd1.Manager', 'JobRemoved', 'uoss',
+                    [job_id, job_path, name, 'done'])
+
+
+def StartUnit(self, name, _mode):
+    job_id = self.next_job_id
+    self.next_job_id += 1
+
+    job_path = '/org/freedesktop/systemd1/Job/%d' % (job_id)
+    GLib.idle_add(lambda: emit_job_new_remove(self, job_id, job_path, name))
```

Comment:
If you want to do this on request, you could also do the async calls way that as done in polkitd

Reply:
I left these as is, what matters is that they are emitted after returning, and this felt like the simplest way to achieve that.

Reply:
So... I've given this again a try, and I'd still prefer to just use pure python-dbus stuff here, and this can be done with just https://github.com/3v1n0/python-dbusmock/commit/dac4db16708a4dc8d52b9912bbc327c9840ed8ea which IMHO is still nicer (and in theory will allow to make it do a delayed if someone will need it in future).